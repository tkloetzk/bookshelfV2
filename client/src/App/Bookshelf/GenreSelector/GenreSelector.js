import React, { useEffect } from 'react'
import Button from '@material-ui/core/Button'
import ButtonGroup from '@material-ui/core/ButtonGroup'
import { makeStyles } from '@material-ui/core/styles'
import Grid from '@material-ui/core/Grid'
import { useDispatch, useSelector } from 'react-redux'
import map from 'lodash/map'
import Select from 'react-select'
import Chip from '@material-ui/core/Chip'
import CancelIcon from '@material-ui/icons/Cancel'
import ToggleButton from '@material-ui/lab/ToggleButton'
import ToggleButtonGroup from '@material-ui/lab/ToggleButtonGroup'
import { SELECTED_GENRES } from '../../../store/bookshelf/bookshelfActionTypes'
import { getGenres } from '../../../store/bookshelf/bookshelfActions'

const useStyles = makeStyles(theme => ({
  container: {
    padding: theme.spacing(1),
    margin: 0,
    width: '100%',
  },
  legend: {
    alignSelf: 'center',
    paddingRight: '10px',
  },
  formGroup: {
    justifyContent: 'center',
  },
}))

function MultiValue(props) {
  return (
    <Chip
      tabIndex={-1}
      label={props.children}
      onDelete={props.removeProps.onClick}
      deleteIcon={<CancelIcon {...props.removeProps} />}
    />
  )
}

const components = {
  MultiValue,
}

export default function GenreSelector({
  setSelector,
  setFilters,
  selector,
  filters,
}) {
  const classes = useStyles()
  const dispatch = useDispatch()
  const genres = useSelector(state => state.bookshelf.genres)
  const [selectedGenres, setSelectedGenres] = React.useState([])

  useEffect(() => {
    dispatch(getGenres())
  }, [dispatch])

  function handleChange(value) {
    setSelectedGenres(value)

    const genreValues = map(value, genre => genre.value)

    dispatch({
      type: SELECTED_GENRES,
      selectedGenres: genreValues,
    })
  }

  return (
    <Grid
      container
      justify="center"
      alignItems="center"
      spacing={4}
      className={classes.container}
    >
      <Grid item xs={2}>
        <ToggleButtonGroup
          onChange={(e, value) => setFilters(value)}
          value={filters}
          aria-label="OWNED/READ genres"
          size="small"
        >
          <ToggleButton value="owned" aria-label="OWNED">
            OWNED
          </ToggleButton>
          <ToggleButton value="read" aria-label="READ">
            READ
          </ToggleButton>
        </ToggleButtonGroup>
      </Grid>
      <Grid item xs={6}>
        <Select
          classes={classes}
          components={components}
          value={selectedGenres}
          onChange={handleChange}
          options={genres}
          isMulti
        />
      </Grid>
      <Grid item xs={2}>
        <ToggleButtonGroup
          onChange={(e, value) => setSelector(value)}
          value={selector}
          aria-label="AND/OR genres"
          exclusive
          size="small"
        >
          <ToggleButton value="OR" aria-label="OR">
            OR
          </ToggleButton>
          <ToggleButton value="AND" aria-label="AND">
            AND
          </ToggleButton>
        </ToggleButtonGroup>
      </Grid>
    </Grid>
  )
}
