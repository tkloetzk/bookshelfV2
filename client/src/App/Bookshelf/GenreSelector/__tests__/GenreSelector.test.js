import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import {
  render,
  fireEvent,
  cleanup,
  wait,
  renderHook,
} from '@testing-library/react'
import { MuiThemeProvider } from '@material-ui/core/styles'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import muiTheme from '../../../../config/themeConfig'
import GenreSelector from '../GenreSelector'

const mockStore = configureMockStore()

describe('GenreSelector', () => {
  let book
  let props
  beforeEach(() => {
    props = {
      setBookshelfFiltered: jest.fn(),
    }
    book = {
      amazonAverageRating: 4.5,
      amazonRatingsCount: 553,
      categories: ['Parenting'],
      description:
        'Answers the most important parenting questions about raising children from birth through kindergarten, covering such topics as discipline, sleeping, day care, safety, independence, and feeding.',
      differences: [
        {
          currentValue: 3.9,
          key: 'amazonAverageRating',
          newValue: 4.5,
        },
      ],
      goodreadsAverageRating: 4,
      goodreadsRatingsCount: 40,
      isbn: '9781402218279',
      price: '',
      subtitle:
        'From Birth to Kindergarten, Answers to the Top 150 Questions about Raising a Young Child',
      thumbnail:
        'http://books.google.com/books/content?id=oD0omQEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api',
      title: 'The New Baby Answer Book',
    }
  })

  describe('render', () => {
    it('should render as expected with no genres', () => {
      const store = mockStore({
        bookshelf: { bookshelf: [], genres: [] },
      })
      store.dispatch = jest.fn()

      const { asFragment } = render(
        <Provider store={store}>
          <MuiThemeProvider theme={muiTheme}>
            <GenreSelector {...props} />
          </MuiThemeProvider>
        </Provider>
      )
      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(asFragment()).toMatchSnapshot()
    })
    it('should render as expected with genres', () => {
      const store = mockStore({
        bookshelf: {
          bookshelf: [],
          genres: ['Parenting', 'Newborn', 'Self-Help'],
        },
      })
      store.dispatch = jest.fn()

      const { asFragment } = render(
        <Provider store={store}>
          <MuiThemeProvider theme={muiTheme}>
            <GenreSelector {...props} />
          </MuiThemeProvider>
        </Provider>
      )
      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(asFragment()).toMatchSnapshot()
    })
  })
  // describe('filters', () => {
  //   it('returns filters genres of an AND selector', async () => {
  //     const store = mockStore({
  //       bookshelf: {
  //         bookshelf: [
  //           book,
  //           {
  //             isbn: '1234',
  //             title: 'Test Title',
  //             categories: ['Parenting', 'Newborn'],
  //           },
  //         ],
  //         genres: ['Parenting', 'Newborn', 'Self-Help'],
  //       },
  //     })
  //     store.dispatch = jest.fn()

  //     const { asFragment, getByTestId } = render(
  //       <Provider store={store}>
  //         <MuiThemeProvider theme={muiTheme}>
  //           <GenreSelector {...props} />
  //         </MuiThemeProvider>
  //       </Provider>
  //     )

  //     expect(wrapper.state('openIndex')).toBe(0)

  // await wait(() => {
  //   fireEvent.change(getByTestId('selectGenres'), {
  //     target: { value: [{ label: 'Religion', value: 'Religion' }] },
  //   })

  //   fireEvent.click(getByTestId('andButton'))
  // })

  // expect(asFragment()).toMatchSnapshot()
  //   })
  // })
  // describe('setBookshelfFiltered', () => {
  // it('is called when bookshelf prop changes', () => {
  //   const store = mockStore({
  //     bookshelf: {
  //       bookshelf: [],
  //       genres: ['Parenting', 'Newborn', 'Self-Help'],
  //     },
  //   })
  //   store.dispatch = jest.fn()

  //   const { asFragment, rerender } = render(
  //     <Provider store={store}>
  //       <MuiThemeProvider theme={muiTheme}>
  //         <GenreSelector {...props} />
  //       </MuiThemeProvider>
  //     </Provider>
  //   )

  //   const store2 = mockStore({
  //     bookshelf: {
  //       bookshelf: [book],
  //       genres: ['Parenting', 'Newborn', 'Self-Help'],
  //     },
  //   })
  //   store2.dispatch = jest.fn()
  //   rerender(
  //     <Provider store={store2}>
  //       <MuiThemeProvider theme={muiTheme}>
  //         <GenreSelector {...props} />
  //       </MuiThemeProvider>
  //     </Provider>
  //   )
  //   expect(props.setBookshelfFiltered).toHaveBeenCalledTimes(0)
  // })
  // })
})
