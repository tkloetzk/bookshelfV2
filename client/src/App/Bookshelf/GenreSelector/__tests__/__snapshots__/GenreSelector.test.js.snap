// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreSelector filters returns filters genres of an AND selector 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-container-1 MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        aria-label="OWNED/READ genres"
        class="MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-label="OWNED"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall MuiToggleButton-sizeSmall"
          tabindex="0"
          type="button"
          value="owned"
        >
          <span
            class="MuiToggleButton-label"
          >
            OWNED
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="READ"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall MuiToggleButton-sizeSmall"
          data-testid=""
          tabindex="0"
          type="button"
          value="read"
        >
          <span
            class="MuiToggleButton-label"
          >
            READ
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class=" css-2b097c-container"
      >
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-1hwfws3"
          >
            <div
              class=" css-1wa3eu0-placeholder"
            >
              Select...
            </div>
            <div
              class="css-w8afj7-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-2-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1g48xl4-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        aria-label="AND/OR genres"
        class="MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-label="OR"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall MuiToggleButton-sizeSmall"
          tabindex="0"
          type="button"
          value="OR"
        >
          <span
            class="MuiToggleButton-label"
          >
            OR
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="AND"
          aria-pressed="true"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall Mui-selected MuiToggleButton-sizeSmall"
          data-testid="andButton"
          tabindex="0"
          type="button"
          value="AND"
        >
          <span
            class="MuiToggleButton-label"
          >
            AND
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GenreSelector render should render as expected with genres 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-container-4 MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        aria-label="OWNED/READ genres"
        class="MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-label="OWNED"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall MuiToggleButton-sizeSmall"
          tabindex="0"
          type="button"
          value="owned"
        >
          <span
            class="MuiToggleButton-label"
          >
            OWNED
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="READ"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall MuiToggleButton-sizeSmall"
          tabindex="0"
          type="button"
          value="read"
        >
          <span
            class="MuiToggleButton-label"
          >
            READ
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class=" css-2b097c-container"
      >
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-1hwfws3"
          >
            <div
              class=" css-1wa3eu0-placeholder"
            >
              Select...
            </div>
            <div
              class="css-w8afj7-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-2-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1g48xl4-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <div
        aria-label="AND/OR genres"
        class="MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-label="OR"
          aria-pressed="true"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall Mui-selected MuiToggleButton-sizeSmall"
          tabindex="0"
          type="button"
          value="OR"
        >
          <span
            class="MuiToggleButton-label"
          >
            OR
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="AND"
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedSizeSmall MuiToggleButton-sizeSmall"
          tabindex="0"
          type="button"
          value="AND"
        >
          <span
            class="MuiToggleButton-label"
          >
            AND
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GenreSelector render should render as expected with no genres 1`] = `<DocumentFragment />`;
